Cesium3DTilesSelection/src/Tileset.cpp

```diff
@@ -294,6 +310,127 @@ Tileset::updateViewOffline(const std::vector<ViewState>& frustums) {
   return this->_updateResult;
 }
 
+bool Tileset::tryCompleteHeightRequest(
+    HeightRequest& request,
+    std::set<Tile*>& tilesNeedingLoading) {
+  Tile* pRoot = _pTilesetContentManager->getRootTile();
+
+  bool tileStillNeedsLoading = false;
+  std::vector<std::string> warnings;
+  for (TerrainQuery& query : request.queries) {
+    query.candidateTiles.clear();
+
+    query.findCandidateTiles(pRoot, warnings);
+
+    // If any candidates need loading, add to return set
+    for (Tile* pTile : query.candidateTiles) {
+      if (pTile->getState() != TileLoadState::Done) {
+        tilesNeedingLoading.insert(pTile);
+        tileStillNeedsLoading = true;
+      }
+    }
+  }
+
+  // Bail if we're waiting on tiles to load
+  if (tileStillNeedsLoading)
+    return false;
+
+  // Do the intersect tests
+  for (TerrainQuery& query : request.queries) {
+    for (Tile* pTile : query.candidateTiles)
+      query.intersectVisibleTile(pTile);
+  }
+
+  // All rays are done, create results
+  Tileset::HeightResults results;
+
+  // Start with any warnings from tile traversal
+  results.warnings = std::move(warnings);
+
+  // Populate results with completed queries
+  for (TerrainQuery& query : request.queries) {
+    Tileset::HeightResults::CoordinateResult coordinateResult = {
+        query.intersectResult.hit.has_value(),
+        std::move(query.inputCoordinate)};
+
+    // Add query warnings into the height result
+    std::copy(
+        query.intersectResult.warnings.begin(),
+        query.intersectResult.warnings.end(),
+        std::back_inserter(results.warnings));
+
+    if (coordinateResult.heightAvailable)
+      coordinateResult.coordinate.height =
+          RAY_ORIGIN_HEIGHT -
+          glm::sqrt(query.intersectResult.hit->rayToWorldPointDistanceSq);
+
+    results.coordinateResults.push_back(coordinateResult);
+  }
+
+  request.promise.resolve(std::move(results));
+  return true;
+}
+
+void Tileset::visitHeightRequests() {
+  if (_heightRequests.size() == 0)
+    return;
+
+  // Go through all requests, either complete them, or gather the tiles they
+  // need for completion
+  std::set<Tile*> tilesNeedingLoading;
+  for (auto it = _heightRequests.begin(); it != _heightRequests.end();) {
+    HeightRequest& request = *it;
+    if (!tryCompleteHeightRequest(request, tilesNeedingLoading)) {
+      ++it;
+    } else {
+      auto deleteIt = it;
+      ++it;
+      _heightRequests.erase(deleteIt);
+    }
+  }
+
+  // Add a load request for tiles that are needed, and haven't started
+  TileLoadPriorityGroup priorityGroup = TileLoadPriorityGroup::Urgent;
+  double priority = 0.0;
+  for (Tile* pTile : tilesNeedingLoading) {
+    TileLoadState loadState = pTile->getState();
+
+    CESIUM_ASSERT(loadState != TileLoadState::Done);
+
+    // Push to appropriate queue based on state
+    if (loadState == TileLoadState::Unloaded) {
```

Comment:
Can we just call `addTileToLoadQueue` here? It seems like we've repeated some of that logic here. I guess the main problem is that `addTileToLoadQueue` assumes (asserts) that the tile is not already in the load queue, whereas here we don't know whether that is true. But that check is moderately expensive, so maybe there's a way we can eliminate it anyway?

Reply:
That would be ideal. Might need to change `addTileToLoadQueue` slightly to make sure it all works. But it's reasonable to consider `addTileToLoadQueue` the official way to do this, whether you're traversing based on a camera, height query, or whatever functionality comes in the future...